modules/sdk-core/src/account-lib/mpc/tss/ecdsa/rangeproof.ts

```diff
@@ -55,13 +55,127 @@ export async function randomCoPrimeTo(x: bigint): Promise<bigint> {
  * @returns {NTilde} The generated NTilde values.
  */
 export async function generateNTilde(bitlength: number): Promise<NTilde> {
-  const ntilde = await generateModulus(bitlength);
+  const { n: ntilde, q1, q2 } = await generateModulus(bitlength);
   const [f1, f2] = await Promise.all([randomCoPrimeTo(ntilde), randomCoPrimeTo(ntilde)]);
   const h1 = modPow(f1, BigInt(2), ntilde);
-  const h2 = modPow(f2, BigInt(2), ntilde);
-  return { ntilde, h1, h2 };
+  const h2 = modPow(h1, f2, ntilde);
+  const beta = modInv(f2, q1 * q2);
+  const [h1wrtH2Proofs, h2wrtH1Proofs] = await Promise.all([
+    generateNTildeProof(
+      {
+        h1: h1,
+        h2: h2,
+        ntilde: ntilde,
+      },
+      f2,
+      q1,
+      q2
+    ),
+    generateNTildeProof(
+      {
+        h1: h2,
+        h2: h1,
+        ntilde: ntilde,
+      },
+      beta,
+      q1,
+      q2
+    ),
+  ]);
+  return {
+    ntilde,
+    h1,
+    h2,
+    ntildeProof: {
+      alpha: h1wrtH2Proofs.alpha.concat(h2wrtH1Proofs.alpha),
+      t: h1wrtH2Proofs.t.concat(h2wrtH1Proofs.t),
+    },
+  };
 }
 
+/**
+ * Generate iterations of Ntilde, h1, h2 discrete log proofs.
+ * @param {NTilde} ntilde Ntilde, h1, h2 to generate the proofs for.
+ * @param {bigint} x Either alpha or beta depending on whether it is a discrete log proof of
+ * h1 w.r.t h2 or h2 w.r.t h1.
+ * @param {bigint} q1 The Sophie Germain prime associated with the first safe prime p1 used to generate Ntilde.
+ * @param {bigint} q2 The Sophie Germain prime associated with the second safe prime p2 used to generate Ntilde.
+ * @param {number} iterations Number of iterations of the ZK proof
+ * (default is 128 as recommened by https://blog.verichains.io/p/vsa-2022-120-multichain-key-extraction)
+ * @returns {NTildeProof} The generated NTilde Proofs.
+ */
+export async function generateNTildeProof(
+  ntilde: NTilde,
+  x: bigint,
+  q1: bigint,
+  q2: bigint,
+  iterations = 128
+): Promise<NtildeProof> {
+  const q1MulQ2 = q1 * q2;
```

Comment:
I'm aware that this is a port of binance's DLN proof https://github.com/bnb-chain/tss-lib/blob/master/crypto/dlnproof/proof.go, where q1, q2 are not the safe primes P, Q but the Sophie Germain primes associated.  CGGMP21 section 6.4 (https://eprint.iacr.org/2021/060.pdf) presents essentially the same proof, but $a_i$ is sampled from $Z_{\phi(N)}$, I'm not sure if it's the equivalent of how it's implemented here.

Reply:
So my understanding of what you shared is that the idea is we want to sample from $Z_{\phi(N)}$ where $\phi(N)$ is the euler function on $N$. It seems that this is basically $(p - 1)(q - 1)$ giving the number of relatively prime integers to N. What we're doing now is sampling from $\frac{(p - 1)(q - 1)}{4}$ since the sophie germains are both $\frac{p - 1}{2}$ and $\frac{q - 1}{2}$. Will look into why that might be equivalent.

Reply:
Will put on the list to ask cryptographer.